include/crocoddyl/core/constraints/residual.hxx

```diff
@@ -0,0 +1,301 @@
+///////////////////////////////////////////////////////////////////////////////
+
+#include "crocoddyl/core/utils/exception.hpp"
+
+namespace crocoddyl {
+
+template <typename Scalar>
+ConstraintModelResidualTpl<Scalar>::ConstraintModelResidualTpl(boost::shared_ptr<typename Base::StateAbstract> state,
+                                                               boost::shared_ptr<ResidualModelAbstract> residual,
+                                                               const VectorXs& lower, const VectorXs& upper)
+    : Base(state, residual,
+           2 * ((upper - lower).array() != 0.).count() -
+               (upper.array() == std::numeric_limits<Scalar>::infinity()).count() -
+               (lower.array() == -std::numeric_limits<Scalar>::infinity()).count(),
+           ((upper - lower).array() == 0.).count()),
+      lb_(lower),
+      ub_(upper),
+      constraint_type_(residual->get_nr()) {
+  for (std::size_t i = 0; i < residual_->get_nr(); ++i) {
+    if (isfinite(lb_(i)) && isfinite(ub_(i))) {
+      if (lb_(i) - ub_(i) > 0) {
+        throw_pretty("Invalid argument: the upper bound is not equals / higher than the lower bound.")
+      }
+    }
+  }
+  if ((lb_.array() == std::numeric_limits<Scalar>::infinity()).any()) {
+    throw_pretty("Invalid argument: the lower bound cannot contain a positive infinity value");
+  }
+  if ((ub_.array() == -std::numeric_limits<Scalar>::infinity()).any()) {
+    throw_pretty("Invalid argument: the lower bound cannot contain a negative infinity value");
+  }
+  updateConstraintType();
+}
+
+template <typename Scalar>
+ConstraintModelResidualTpl<Scalar>::ConstraintModelResidualTpl(boost::shared_ptr<typename Base::StateAbstract> state,
+                                                               boost::shared_ptr<ResidualModelAbstract> residual)
+    : Base(state, residual, 0, residual->get_nr()),
+      lb_(VectorXs::Zero(nh_)),
+      ub_(VectorXs::Zero(nh_)),
+      constraint_type_(residual->get_nr()) {
+  constraint_type_ = 0;
+}
+
+template <typename Scalar>
+ConstraintModelResidualTpl<Scalar>::~ConstraintModelResidualTpl() {}
+
+template <typename Scalar>
+void ConstraintModelResidualTpl<Scalar>::calc(const boost::shared_ptr<ConstraintDataAbstract>& data,
+                                              const Eigen::Ref<const VectorXs>& x,
+                                              const Eigen::Ref<const VectorXs>& u) {
+  // Compute the constraint residual
+  residual_->calc(data->residual, x, u);
+
+  // Fill the residual values for its corresponding type of constraint
+  std::size_t nh_i = 0;
+  std::size_t ng_i = 0;
+  const std::size_t nr = residual_->get_nr();
+  for (std::size_t i = 0; i < nr; ++i) {
+    if (constraint_type_(i) == 0) {  // equality constraint
+      data->h(nh_i) = data->residual->r(i) - ub_(i);
+      ++nh_i;
+    } else if (constraint_type_(i) == 1) {  // inequality constraint
+      data->g(ng_i) = data->residual->r(i) - ub_(i);
```

Comment:
This is a bad strategy to loose the information on the double bounds. Better store both for later use by the solver.

Reply:
To pass the bound from the constraint residual to the solver, we need to perform a number of copies. These copies are from the constraint residual to the constraint manager, then from the constraint manager to the diff action model, and then from the diff action to the action model. This is very convolute and error-prone.

Additionally, we need to do this every time as we wish to enable users to update the bounds. There are many examples where we wish this.

One might share the memory between these objects, however, this is not trivial as the bounds vector for the diff action (or action) model bounds depends on a number of constraint residual bounds. Moreover, some constraint residuals might be equality or inequality, and we are not arranging them using this property.

Again, this is very convolute and painful to develop (simpler equals better). Could you motivate the need of passing this information to the solver?

Reply:
The solver would benefit from knowing that there are some double bounds, and would handle them more efficiently than considering two simple bounds. 

There is no complexity at all in handling that. You just need to consider double bounds as one of the possible option for the high-level action-model API. Again, there are several ways of doing it: either allow double bounds by default with infinite values, or with additional flags to activate one side of the double bounds, or store double bounds apart from single bounds. 

At the end, you are storing this information in the specialized (pinocchio) part of your front end. So the point here is simply to consider it at the low level.

Reply:
Defined the bounds in the front-end